# ============================================================================
# Aureus Sigma Capital - Risk Overlay Analysis
# Notebook: 06_risk_overlay_analysis.ipynb
#
# Description:
# Phase 8 Risk Management Implementation - Testing three distinct risk overlay
# mechanisms to reduce maximum drawdown from -45.2% to institutional target
# <25% while preserving validated alpha generation (Sharpe >1.2).
#
# Author: Duc Nguyen, Quantitative Finance Expert
# Date: July 27, 2025
# Version: 1.0 - Institutional Risk Management Framework
# ============================================================================

In [1]:
# Core imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import seaborn as sns
from datetime import datetime
import pickle
import warnings
from scipy import stats
from typing import Dict, List, Tuple, Optional
import yaml
from sqlalchemy import create_engine
from pathlib import Path
import sys

warnings.filterwarnings('ignore')

# --- INSTITUTIONAL PALETTE (Blackstone-inspired) ---
FACTOR_COLORS = {
    'Strategy': '#16A085',          # Blackstone Teal (primary)
    'Benchmark': '#34495E',         # Warm charcoal (secondary)
    'Positive': '#27AE60',         # Professional green
    'Negative': '#C0392B',         # Sophisticated red
    'Drawdown': '#E67E22',         # Sophisticated orange
    'Sharpe': '#2980B9',           # Institutional blue
    'Grid': '#BDC3C7',
    'Text_Primary': '#2C3E50',
    'Neutral': '#7F8C8D',
    # Risk overlay colors
    'Regime': '#9B59B6',           # Purple for regime overlay
    'VolTarget': '#E74C3C',        # Red for volatility targeting
    'DynReversal': '#F39C12',      # Orange for dynamic reversal
    'Control': '#2C3E50'           # Dark for control baseline
}

GRADIENT_PALETTES = {
    'performance': ['#C0392B', '#FFFFFF', '#27AE60'],  # Red-White-Green
}

# --- ENHANCED VISUALIZATION CONFIGURATION ---
plt.style.use('default')
plt.rcParams.update({
    'figure.dpi': 300, 'savefig.dpi': 300, 'figure.figsize': (15, 8),
    'figure.facecolor': 'white', 'font.size': 11,
    'axes.facecolor': 'white', 'axes.edgecolor': FACTOR_COLORS['Text_Primary'],
    'axes.linewidth': 1.0, 'axes.grid': True, 'axes.axisbelow': True,
    'axes.labelcolor': FACTOR_COLORS['Text_Primary'], 'axes.titlesize': 14,
    'axes.titleweight': 'bold', 'axes.titlecolor': FACTOR_COLORS['Text_Primary'],
    'grid.color': FACTOR_COLORS['Grid'], 'grid.alpha': 0.3, 'grid.linewidth': 0.5,
    'legend.frameon': False, 'legend.fontsize': 10,
    'xtick.color': FACTOR_COLORS['Text_Primary'], 'ytick.color': FACTOR_COLORS['Text_Primary'],
    'xtick.labelsize': 10, 'ytick.labelsize': 10,
    'lines.linewidth': 2.0, 'lines.solid_capstyle': 'round'
})

print("📊 Visualization environment configured with institutional palette.")

print("\n" + "=" * 70)
print("🎯 Aureus Sigma: Phase 8 Risk Overlay Analysis")
print(f"   Version: 1.0 - Date: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
print("=" * 70)
print("\n🎯 Phase 8 Mission:")
print("   • Reduce maximum drawdown from -45.2% to institutional target <25%")
print("   • Preserve validated alpha generation (maintain Sharpe >1.2)")
print("   • Test 3 risk overlay mechanisms on quarterly rebalancing baseline")
print("   • Optimize for Calmar Ratio (Annual Return / Max Drawdown)")
print("-" * 70)

📊 Visualization environment configured with institutional palette.

🎯 Aureus Sigma: Phase 8 Risk Overlay Analysis
   Version: 1.0 - Date: 2025-07-27 09:07:19

🎯 Phase 8 Mission:
   • Reduce maximum drawdown from -45.2% to institutional target <25%
   • Preserve validated alpha generation (maintain Sharpe >1.2)
   • Test 3 risk overlay mechanisms on quarterly rebalancing baseline
   • Optimize for Calmar Ratio (Annual Return / Max Drawdown)
----------------------------------------------------------------------


## 1. Load Core Data and Establish Project Context

In [2]:
# ============================================================================
# CELL 2: LOAD PHASE 7 RESULTS AND PROJECT DATA
# ============================================================================

# Establish project paths
project_root = Path.cwd()
while not (project_root / 'production').exists() and not (project_root / 'config').exists():
    if project_root.parent == project_root:
        raise FileNotFoundError("Could not find project root")
    project_root = project_root.parent

phase7_path = project_root / "production" / "tests" / "phase7_institutional_backtesting"
phase8_path = project_root / "production" / "tests" / "phase8_risk_management"

print("📂 Loading Phase 7 validated results and core data...")
print(f"   Phase 7 path: {phase7_path}")
print(f"   Phase 8 path: {phase8_path}")

# Load Phase 7 canonical backtest results
with open(phase7_path / "canonical_backtest_results.pkl", "rb") as f:
    phase7_results = pickle.load(f)

# Load core data objects from Phase 7
with open(phase7_path / "factor_data.pkl", "rb") as f:
    factor_data_obj = pickle.load(f)
with open(phase7_path / "daily_returns.pkl", "rb") as f:
    returns_data_obj = pickle.load(f)
with open(phase7_path / "benchmark_returns.pkl", "rb") as f:
    benchmark_data_obj = pickle.load(f)

# Extract data
factor_data = factor_data_obj['data']
daily_returns = returns_data_obj['data']
benchmark_returns = benchmark_data_obj['data']

print("✅ Phase 7 validated results loaded:")
print(f"   Monthly strategy performance: {(phase7_results['net_returns'].mean() * 252):.2%} annual return")
print(f"   Monthly strategy Sharpe: {phase7_results['performance_summary']['sharpe_ratio']:.2f}")
print(f"   Maximum drawdown (monthly): {phase7_results['performance_summary']['annual_vol']:.2%}")

# Load sector mappings
print("\n🏗️ Loading sector information...")
config_path = project_root / 'config' / 'database.yml'
with open(config_path, 'r') as f:
    db_config = yaml.safe_load(f)['production']

engine = create_engine(
    f"mysql+pymysql://{db_config['username']}:{db_config['password']}@"
    f"{db_config['host']}/{db_config['schema_name']}"
)

sector_info = pd.read_sql("SELECT ticker, sector FROM master_info WHERE sector IS NOT NULL", engine)
sector_info = sector_info.drop_duplicates(subset=['ticker']).set_index('ticker')
engine.dispose()

print(f"✅ Loaded sector mappings for {len(sector_info)} tickers")

# Align data for Phase 8 analysis
common_index = factor_data.index.intersection(daily_returns.index).intersection(benchmark_returns.index)
common_tickers = factor_data.columns.get_level_values(1).intersection(daily_returns.columns).unique().intersection(sector_info.index)

# Extract QVM scores and align all data
qvm_scores = factor_data.loc[common_index, ('qvm_composite_score', common_tickers)]
qvm_scores.columns = qvm_scores.columns.droplevel(0)
daily_returns_aligned = daily_returns.loc[common_index, common_tickers]
benchmark_returns_aligned = benchmark_returns.loc[common_index]

print("\n🔗 Data aligned for Phase 8 analysis:")
print(f"   Date range: {common_index.min().date()} to {common_index.max().date()}")
print(f"   Trading days: {len(common_index)}")
print(f"   Universe size: {len(common_tickers)} stocks")
print(f"   QVM scores shape: {qvm_scores.shape}")

📂 Loading Phase 7 validated results and core data...
   Phase 7 path: /Users/ducnguyen/Library/CloudStorage/GoogleDrive-duc.nguyentcb@gmail.com/My Drive/quant-world-invest/factor_investing_project/production/tests/phase7_institutional_backtesting
   Phase 8 path: /Users/ducnguyen/Library/CloudStorage/GoogleDrive-duc.nguyentcb@gmail.com/My Drive/quant-world-invest/factor_investing_project/production/tests/phase8_risk_management
✅ Phase 7 validated results loaded:
   Monthly strategy performance: 18.97% annual return
   Monthly strategy Sharpe: 1.52
   Maximum drawdown (monthly): 13.03%

🏗️ Loading sector information...
✅ Loaded sector mappings for 728 tickers

🔗 Data aligned for Phase 8 analysis:
   Date range: 2016-01-05 to 2025-07-25
   Trading days: 2381
   Universe size: 714 stocks
   QVM scores shape: (2381, 714)


## 2. Market Regime Identification (from Phase 7)

Load and validate the market regime framework that will be used for risk overlays.

In [4]:
# ============================================================================
# CELL 3: MARKET REGIME FRAMEWORK (FROM PHASE 7 ATTRIBUTION ANALYSIS)
# ============================================================================

def identify_market_regimes(benchmark_returns: pd.Series, 
                          bear_threshold: float = -0.20,
                          vol_window: int = 60,
                          trend_window: int = 200) -> pd.DataFrame:
    """
    Identifies market regimes using multiple criteria (from Phase 7 validation):
    - Bear: Drawdown > 20% from peak
    - Stress: Rolling volatility in top quartile
    - Bull: Price above trend MA and not Bear/Stress
    - Sideways: Everything else
    """
    print("🔍 Identifying market regimes using Phase 7 validated methodology...")
    
    # Calculate cumulative returns and drawdowns
    cumulative = (1 + benchmark_returns).cumprod()
    drawdown = (cumulative / cumulative.cummax() - 1)
    
    # 1. Bear Market Regime
    is_bear = drawdown < bear_threshold
    
    # 2. High-Stress Regime (rolling volatility)
    rolling_vol = benchmark_returns.rolling(vol_window).std() * np.sqrt(252)
    vol_75th = rolling_vol.quantile(0.75)
    is_stress = rolling_vol > vol_75th
    
    # 3. Bull/Sideways (trend-based)
    trend_ma = cumulative.rolling(trend_window).mean()
    is_above_trend = cumulative > trend_ma
    
    # Combine into regime classification
    regimes = pd.DataFrame(index=benchmark_returns.index)
    regimes['is_bear'] = is_bear
    regimes['is_stress'] = is_stress
    regimes['is_bull'] = is_above_trend & ~is_bear & ~is_stress
    regimes['is_sideways'] = ~is_above_trend & ~is_bear & ~is_stress
    
    # Create primary regime classification
    regimes['regime'] = 'Undefined'
    regimes.loc[regimes['is_bear'], 'regime'] = 'Bear'
    regimes.loc[regimes['is_stress'] & ~regimes['is_bear'], 'regime'] = 'Stress'
    regimes.loc[regimes['is_bull'], 'regime'] = 'Bull'
    regimes.loc[regimes['is_sideways'], 'regime'] = 'Sideways'
    
    # Summary statistics
    regime_counts = regimes['regime'].value_counts()
    regime_pcts = (regime_counts / len(regimes)) * 100
    
    print("\n📊 Regime Distribution (Phase 8 Implementation):")
    for regime, pct in regime_pcts.items():
        days = regime_counts[regime]
        print(f"   {regime:10s}: {days:5d} days ({pct:5.1f}%)")
    
    # Add additional metrics
    regimes['drawdown'] = drawdown
    regimes['rolling_vol'] = rolling_vol
    regimes['cumulative_return'] = cumulative
    
    return regimes

# Execute regime identification
market_regimes = identify_market_regimes(benchmark_returns_aligned)

# Validate regime signals for risk overlay implementation
bear_stress_days = (market_regimes['regime'].isin(['Bear', 'Stress'])).sum()
total_days = len(market_regimes)
risk_coverage = bear_stress_days / total_days

print(f"\n🎯 Risk Overlay Coverage Analysis:")
print(f"   Bear + Stress periods: {bear_stress_days:,} days ({risk_coverage:.1%})")
print(f"   Risk reduction opportunities: {risk_coverage:.1%} of trading days")

if risk_coverage > 0.35:
    print("   ✅ GOOD COVERAGE: Sufficient Bear/Stress periods for risk overlay testing")
elif risk_coverage > 0.20:
    print("   ⚠️ MODERATE COVERAGE: Limited but adequate risk periods")
else:
    print("   ❌ LOW COVERAGE: May not provide sufficient risk reduction opportunities")

print(f"\n✅ Market regime framework ready for Phase 8 implementation")

🔍 Identifying market regimes using Phase 7 validated methodology...

📊 Regime Distribution (Phase 8 Implementation):
   Bull      :  1004 days ( 42.2%)
   Bear      :   768 days ( 32.3%)
   Sideways  :   335 days ( 14.1%)
   Stress    :   274 days ( 11.5%)

🎯 Risk Overlay Coverage Analysis:
   Bear + Stress periods: 1,042 days (43.8%)
   Risk reduction opportunities: 43.8% of trading days
   ✅ GOOD COVERAGE: Sufficient Bear/Stress periods for risk overlay testing

✅ Market regime framework ready for Phase 8 implementation


## 3. Quarterly Baseline Implementation (Control Group)

Establish the quarterly rebalancing baseline that will serve as our control group for comparing risk overlay mechanisms.

In [5]:
# ============================================================================
# CELL 4: QUARTERLY BASELINE IMPLEMENTATION (CONTROL GROUP)
# ============================================================================

# Enhanced strategy configuration for Phase 8
QUARTERLY_BASELINE_CONFIG = {
    "backtest_start_date": "2016-01-01",
    "backtest_end_date": "2025-07-25",
    "selection_percentile": 0.20,
    "rebalance_freq": 'Q',  # QUARTERLY - Phase 8 baseline
    "long_only": True,
    "max_sector_weight": 0.40,
    "max_position_weight": 0.05,
    "transaction_cost_bps": 30
}

print("🚀 Implementing Quarterly Baseline (Control Group)...")
print("\n--- QUARTERLY BASELINE CONFIGURATION ---")
for key, value in QUARTERLY_BASELINE_CONFIG.items():
    print(f"{key:<25}: {value}")

# Load constrained portfolio construction and backtesting engine from Phase 7
def construct_constrained_portfolio(
    factor_scores: pd.Series, 
    sector_info: pd.DataFrame, 
    config: dict
) -> pd.DataFrame:
    """
    Constructs a single, constrained portfolio for a given rebalance date.
    (Validated in Phase 7 - maintaining same logic)
    """
    if factor_scores.empty:
        return pd.DataFrame(columns=['weight', 'sector'])

    # Select top quintile of stocks
    top_quintile_cutoff = factor_scores.quantile(1 - config['selection_percentile'])
    selected_stocks_df = factor_scores[factor_scores >= top_quintile_cutoff].to_frame('factor_score')
    
    # Merge with sector information
    portfolio_df = selected_stocks_df.join(sector_info)
    
    # Handle potential missing sectors after join
    if portfolio_df['sector'].isnull().any():
        portfolio_df.dropna(subset=['sector'], inplace=True)

    if portfolio_df.empty:
        return pd.DataFrame(columns=['weight', 'sector'])

    # Apply sector constraints
    sector_counts = portfolio_df['sector'].value_counts()
    max_stocks_in_portfolio = len(portfolio_df)
    max_stocks_per_sector = int(max_stocks_in_portfolio * config['max_sector_weight'])
    
    final_tickers = set()
    for sector, count in sector_counts.items():
        sector_stocks = portfolio_df[portfolio_df['sector'] == sector]
        if count > max_stocks_per_sector and max_stocks_per_sector > 0:
            top_in_sector = sector_stocks.nlargest(max_stocks_per_sector, 'factor_score').index
            final_tickers.update(top_in_sector)
        else:
            final_tickers.update(sector_stocks.index)
            
    final_portfolio = portfolio_df.loc[list(final_tickers)].copy()
    
    # Assign equal weights
    num_stocks = len(final_portfolio)
    if num_stocks > 0:
        final_portfolio['weight'] = 1.0 / num_stocks
    else:
        return pd.DataFrame(columns=['weight', 'sector'])
        
    return final_portfolio[['weight', 'sector']]


def run_quarterly_baseline_backtest(
    qvm_scores: pd.DataFrame,
    daily_returns: pd.DataFrame,
    sector_info: pd.DataFrame,
    config: dict
) -> Tuple[pd.Series, pd.DataFrame]:
    """
    Runs the quarterly baseline backtest (Control Group for Phase 8).
    """
    print("🚀 Running Quarterly Baseline Backtest (Control Group)...")
    
    # 1. IDENTIFY QUARTERLY REBALANCE DATES
    ideal_rebalance_dates = pd.date_range(
        start=qvm_scores.index.min(), 
        end=qvm_scores.index.max(), 
        freq=config['rebalance_freq']
    )
    print(f"   - Identified {len(ideal_rebalance_dates)} quarterly rebalance dates.")

    # 2. Construct Daily Holdings Matrix
    daily_holdings = pd.DataFrame(index=daily_returns.index, columns=daily_returns.columns).fillna(0.0)
    
    factor_scores_on_rebal_dates = qvm_scores.reindex(ideal_rebalance_dates, method='ffill')

    for i in range(len(factor_scores_on_rebal_dates.index)):
        rebal_date = factor_scores_on_rebal_dates.index[i]
        
        try:
            next_rebal_date = factor_scores_on_rebal_dates.index[i+1]
        except IndexError:
            next_rebal_date = daily_returns.index[-1] + pd.Timedelta(days=1)

        factor_scores_at_rebal = factor_scores_on_rebal_dates.loc[rebal_date].dropna()
        
        if len(factor_scores_at_rebal) > 20:
            portfolio_df = construct_constrained_portfolio(factor_scores_at_rebal, sector_info, config)
            
            if not portfolio_df.empty:
                # Define the holding period for this portfolio
                relevant_days = daily_returns.index[(daily_returns.index > rebal_date) & (daily_returns.index < next_rebal_date)]
                
                if not relevant_days.empty:
                    # Assign the weight Series to each relevant day
                    for day in relevant_days:
                        valid_tickers = portfolio_df.index.intersection(daily_holdings.columns)
                        daily_holdings.loc[day, valid_tickers] = portfolio_df.loc[valid_tickers, 'weight']

    print("   - Constructed quarterly holdings matrix.")

    # 3. PREVENT LOOK-AHEAD BIAS
    daily_holdings_shifted = daily_holdings.shift(1).fillna(0)
    print("   - Shifted holdings by 1 day to prevent look-ahead bias.")

    # 4. CALCULATE GROSS PORTFOLIO RETURNS
    gross_returns = (daily_holdings_shifted * daily_returns).sum(axis=1)
    print("   - Calculated daily gross returns.")

    # 5. MODEL TRANSACTION COSTS
    turnover = (daily_holdings_shifted - daily_holdings_shifted.shift(1)).abs().sum(axis=1) / 2
    transaction_costs = turnover * (config['transaction_cost_bps'] / 10000)
    
    net_returns = gross_returns - transaction_costs
    print("   - Applied transaction costs to get net returns.")
    
    backtest_log = pd.DataFrame({
        'gross_return': gross_returns,
        'net_return': net_returns,
        'turnover': turnover,
        'transaction_cost': transaction_costs,
        'positions': (daily_holdings_shifted > 0).sum(axis=1)
    })

    print("✅ Quarterly baseline backtest complete.")
    return net_returns, backtest_log, daily_holdings_shifted

# Execute quarterly baseline backtest
quarterly_returns, quarterly_log, quarterly_holdings = run_quarterly_baseline_backtest(
    qvm_scores=qvm_scores,
    daily_returns=daily_returns_aligned,
    sector_info=sector_info,
    config=QUARTERLY_BASELINE_CONFIG
)

🚀 Implementing Quarterly Baseline (Control Group)...

--- QUARTERLY BASELINE CONFIGURATION ---
backtest_start_date      : 2016-01-01
backtest_end_date        : 2025-07-25
selection_percentile     : 0.2
rebalance_freq           : Q
long_only                : True
max_sector_weight        : 0.4
max_position_weight      : 0.05
transaction_cost_bps     : 30
🚀 Running Quarterly Baseline Backtest (Control Group)...
   - Identified 38 quarterly rebalance dates.
   - Constructed quarterly holdings matrix.
   - Shifted holdings by 1 day to prevent look-ahead bias.
   - Calculated daily gross returns.
   - Applied transaction costs to get net returns.
✅ Quarterly baseline backtest complete.


In [6]:
# ============================================================================
# QUARTERLY BASELINE PERFORMANCE VALIDATION
# ============================================================================

def calculate_performance_metrics(returns: pd.Series, 
                                  benchmark: pd.Series = None,
                                  risk_free_rate: float = 0.0) -> Dict[str, float]:
    """
    Calculate comprehensive performance metrics for Phase 8 analysis.
    """
    # Basic metrics
    total_return = (1 + returns).prod() - 1
    n_years = len(returns) / 252
    annual_return = (1 + total_return) ** (1 / n_years) - 1 if n_years > 0 else 0
    annual_vol = returns.std() * np.sqrt(252)
    sharpe_ratio = (annual_return - risk_free_rate) / annual_vol if annual_vol > 0 else 0
    
    # Drawdown analysis
    cumulative = (1 + returns).cumprod()
    drawdown = (cumulative / cumulative.cummax() - 1)
    max_drawdown = drawdown.min()
    
    # Calmar Ratio (key metric for Phase 8)
    calmar_ratio = annual_return / abs(max_drawdown) if max_drawdown < 0 else 0
    
    # Downside metrics
    downside_returns = returns[returns < 0]
    downside_vol = downside_returns.std() * np.sqrt(252) if len(downside_returns) > 0 else 0
    sortino_ratio = (annual_return - risk_free_rate) / downside_vol if downside_vol > 0 else 0
    
    # Win rates
    win_rate = (returns > 0).mean()
    
    metrics = {
        'Annual Return (%)': annual_return * 100,
        'Annual Volatility (%)': annual_vol * 100,
        'Sharpe Ratio': sharpe_ratio,
        'Max Drawdown (%)': max_drawdown * 100,
        'Calmar Ratio': calmar_ratio,
        'Sortino Ratio': sortino_ratio,
        'Win Rate (%)': win_rate * 100,
        'Total Days': len(returns)
    }
    
    # Add benchmark comparison if provided
    if benchmark is not None:
        common_idx = returns.index.intersection(benchmark.index)
        excess_returns = returns.loc[common_idx] - benchmark.loc[common_idx]
        tracking_error = excess_returns.std() * np.sqrt(252)
        information_ratio = (excess_returns.mean() * 252) / tracking_error if tracking_error > 0 else 0
        
        metrics['Information Ratio'] = information_ratio
        metrics['Tracking Error (%)'] = tracking_error * 100
    
    return metrics

# Calculate quarterly baseline performance
quarterly_metrics = calculate_performance_metrics(quarterly_returns, benchmark_returns_aligned)

print("\n" + "=" * 70)
print("📊 QUARTERLY BASELINE PERFORMANCE VALIDATION (CONTROL GROUP)")
print("=" * 70)

for metric, value in quarterly_metrics.items():
    if 'Ratio' in metric or 'Rate' in metric:
        print(f"{metric:<25}: {value:8.2f}")
    else:
        print(f"{metric:<25}: {value:8.2f}")

# Validate against Phase 8 targets
print("\n🎯 Phase 8 Target Validation:")
sharpe_target = quarterly_metrics['Sharpe Ratio'] >= 1.2
drawdown_concern = quarterly_metrics['Max Drawdown (%)'] < -25

print(f"   Sharpe Ratio Target (>1.2):  {'✅ PASS' if sharpe_target else '❌ FAIL'} ({quarterly_metrics['Sharpe Ratio']:.2f})")
print(f"   Drawdown Concern (<-25%):     {'⚠️ YES' if drawdown_concern else '✅ NO'} ({quarterly_metrics['Max Drawdown (%)']:.1f}%)")

if drawdown_concern:
    print("\n⚠️ RISK MANAGEMENT REQUIRED: Quarterly baseline exceeds -25% drawdown target")
    print("   Risk overlays are essential for institutional deployment")
else:
    print("\n✅ BASELINE ACCEPTABLE: Quarterly baseline meets drawdown requirements")
    print("   Risk overlays will focus on further optimization")

# Calculate improvement needed
current_calmar = quarterly_metrics['Calmar Ratio']
print(f"\n📈 Baseline Calmar Ratio: {current_calmar:.2f}")
print(f"   Target: Maximize Calmar while maintaining Sharpe >1.2")

print("\n✅ Quarterly baseline established as Control Group for Phase 8 risk overlay testing")


📊 QUARTERLY BASELINE PERFORMANCE VALIDATION (CONTROL GROUP)
Annual Return (%)        :    18.70
Annual Volatility (%)    :    13.32
Sharpe Ratio             :     1.40
Max Drawdown (%)         :   -49.22
Calmar Ratio             :     0.38
Sortino Ratio            :     1.45
Win Rate (%)             :    59.01
Total Days               :  2381.00
Information Ratio        :     0.54
Tracking Error (%)       :    11.48

🎯 Phase 8 Target Validation:
   Sharpe Ratio Target (>1.2):  ✅ PASS (1.40)
   Drawdown Concern (<-25%):     ⚠️ YES (-49.2%)

⚠️ RISK MANAGEMENT REQUIRED: Quarterly baseline exceeds -25% drawdown target
   Risk overlays are essential for institutional deployment

📈 Baseline Calmar Ratio: 0.38
   Target: Maximize Calmar while maintaining Sharpe >1.2

✅ Quarterly baseline established as Control Group for Phase 8 risk overlay testing


In [7]:
# =================================================================
# CELL 5: MARKET REGIME OVERLAY IMPLEMENTATION (TEST A)
# =================================================================

def apply_market_regime_overlay(
    baseline_holdings: pd.DataFrame,
    market_regimes: pd.DataFrame,
    risk_reduction_factor: float = 0.5
) -> pd.DataFrame:
    """
    Apply market regime overlay by reducing exposure during Bear/Stress periods.
    
    Parameters:
    - baseline_holdings: Daily portfolio weights from quarterly baseline
    - market_regimes: Regime classification from Phase 7 framework
    - risk_reduction_factor: Multiplier for Bear/Stress periods (0.5 = 50% exposure)
    
    Returns:
    - risk_managed_holdings: Adjusted portfolio weights with regime overlay
    """
    print(f"🔧 Implementing Market Regime Overlay (Test A)...")
    print(f"    Risk reduction factor: {risk_reduction_factor} (50% exposure during Bear/Stress)")

    # Create a copy of baseline holdings
    risk_managed_holdings = baseline_holdings.copy()

    # Identify Bear and Stress periods
    bear_stress_mask = market_regimes['regime'].isin(['Bear', 'Stress'])
    bear_stress_dates = market_regimes[bear_stress_mask].index

    print(f"    Bear/Stress periods identified: {len(bear_stress_dates):,} days ({len(bear_stress_dates)/len(market_regimes):.1%})")

    # Apply risk reduction to Bear/Stress periods
    common_dates = baseline_holdings.index.intersection(bear_stress_dates)
    if len(common_dates) > 0:
        risk_managed_holdings.loc[common_dates] *= risk_reduction_factor
        print(f"    Applied {risk_reduction_factor} exposure multiplier to {len(common_dates):,} trading days")

    # Calculate portfolio statistics
    normal_exposure = baseline_holdings[~baseline_holdings.index.isin(bear_stress_dates)].sum(axis=1).mean()
    reduced_exposure = risk_managed_holdings[risk_managed_holdings.index.isin(bear_stress_dates)].sum(axis=1).mean()

    print(f"    Average exposure during normal periods: {normal_exposure:.1%}")
    print(f"    Average exposure during Bear/Stress: {reduced_exposure:.1%}")

    return risk_managed_holdings


def run_regime_overlay_backtest(
    regime_holdings: pd.DataFrame,
    daily_returns: pd.DataFrame,
    config: dict,
    overlay_name: str = "Regime Overlay"
) -> Tuple[pd.Series, pd.DataFrame]:
    """
    Run backtest with regime-adjusted holdings.
    """
    print(f"🚀 Running {overlay_name} backtest...")

    # Ensure no look-ahead bias (holdings should already be shifted)
    # But apply one more shift to be absolutely certain
    regime_holdings_shifted = regime_holdings.shift(1).fillna(0)

    # Calculate gross returns
    gross_returns = (regime_holdings_shifted * daily_returns).sum(axis=1)

    # Calculate turnover for transaction costs
    turnover = (regime_holdings_shifted - regime_holdings_shifted.shift(1)).abs().sum(axis=1) / 2
    transaction_costs = turnover * (config['transaction_cost_bps'] / 10000)

    # Net returns
    net_returns = gross_returns - transaction_costs

    # Backtest log
    backtest_log = pd.DataFrame({
        'gross_return': gross_returns,
        'net_return': net_returns,
        'turnover': turnover,
        'transaction_cost': transaction_costs,
        'positions': (regime_holdings_shifted > 0).sum(axis=1),
        'total_exposure': regime_holdings_shifted.sum(axis=1)
    })

    print(f"✅ {overlay_name} backtest complete.")
    return net_returns, backtest_log


# Execute Market Regime Overlay (Test A)
print("\n" + "=" * 70)
print("🎯 IMPLEMENTING TEST A: MARKET REGIME OVERLAY")
print("=" * 70)

# Apply regime overlay to quarterly baseline holdings
regime_overlay_holdings = apply_market_regime_overlay(
    baseline_holdings=quarterly_holdings,
    market_regimes=market_regimes,
    risk_reduction_factor=0.5  # 50% exposure during Bear/Stress
)

# Run backtest with regime overlay
regime_overlay_returns, regime_overlay_log = run_regime_overlay_backtest(
    regime_holdings=regime_overlay_holdings,
    daily_returns=daily_returns_aligned,
    config=QUARTERLY_BASELINE_CONFIG,
    overlay_name="Market Regime Overlay"
)

# Calculate performance metrics for Test A
regime_overlay_metrics = calculate_performance_metrics(regime_overlay_returns, benchmark_returns_aligned)

print("\n📊 MARKET REGIME OVERLAY PERFORMANCE (TEST A):")
print("=" * 50)
for metric, value in regime_overlay_metrics.items():
    if 'Ratio' in metric or 'Rate' in metric:
        print(f"{metric:<25}: {value:8.2f}")
    else:
        print(f"{metric:<25}: {value:8.2f}")

# Compare against baseline
print("\n🔍 TEST A vs BASELINE COMPARISON:")
print("=" * 40)
print(f"{'Metric':<25} {'Baseline':<10} {'Test A':<10} {'Change':<10}")
print("-" * 55)

key_metrics = ['Annual Return (%)', 'Sharpe Ratio', 'Max Drawdown (%)', 'Calmar Ratio']
for metric in key_metrics:
    baseline_val = quarterly_metrics[metric]
    overlay_val = regime_overlay_metrics[metric]
    change = overlay_val - baseline_val
    change_str = f"{change:+.2f}"
    print(f"{metric:<25} {baseline_val:8.2f}    {overlay_val:8.2f}    {change_str:>8s}")

# Risk assessment for Test A
print("\n🎯 TEST A RISK ASSESSMENT:")
print("=" * 30)

sharpe_preserved = regime_overlay_metrics['Sharpe Ratio'] >= 1.2
drawdown_improved = regime_overlay_metrics['Max Drawdown (%)'] > quarterly_metrics['Max Drawdown (%)']
calmar_improved = regime_overlay_metrics['Calmar Ratio'] > quarterly_metrics['Calmar Ratio']
target_achieved = regime_overlay_metrics['Max Drawdown (%)'] > -25.0

print(f"Sharpe Preservation (>1.2):      {'✅ PASS' if sharpe_preserved else '❌ FAIL'} ({regime_overlay_metrics['Sharpe Ratio']:.2f})")
print(f"Drawdown Improvement:            {'✅ YES' if drawdown_improved else '❌ NO'} ({regime_overlay_metrics['Max Drawdown (%)']:.1f}% vs {quarterly_metrics['Max Drawdown (%)']:.1f}%)")
print(f"Calmar Improvement:              {'✅ YES' if calmar_improved else '❌ NO'} ({regime_overlay_metrics['Calmar Ratio']:.2f} vs {quarterly_metrics['Calmar Ratio']:.2f})")
print(f"Target Achieved (<-25%):         {'✅ YES' if target_achieved else '❌ NO'} ({regime_overlay_metrics['Max Drawdown (%)']:.1f}%)")

if target_achieved and sharpe_preserved:
    print("\n🎉 TEST A SUCCESS: Market Regime Overlay achieves institutional targets!")
elif drawdown_improved and sharpe_preserved:
    print("\n✅ TEST A PROGRESS: Significant improvement, may need fine-tuning")
else:
    print("\n⚠️ TEST A MIXED: Review trade-offs between risk reduction and alpha preservation")

print("\n✅ Market Regime Overlay (Test A) analysis complete")


🎯 IMPLEMENTING TEST A: MARKET REGIME OVERLAY
🔧 Implementing Market Regime Overlay (Test A)...
    Risk reduction factor: 0.5 (50% exposure during Bear/Stress)
    Bear/Stress periods identified: 1,042 days (43.8%)
    Applied 0.5 exposure multiplier to 1,042 trading days
    Average exposure during normal periods: 94.5%
    Average exposure during Bear/Stress: 49.6%
🚀 Running Market Regime Overlay backtest...
✅ Market Regime Overlay backtest complete.

📊 MARKET REGIME OVERLAY PERFORMANCE (TEST A):
Annual Return (%)        :    12.83
Annual Volatility (%)    :    10.30
Sharpe Ratio             :     1.25
Max Drawdown (%)         :   -39.58
Calmar Ratio             :     0.32
Sortino Ratio            :     1.22
Win Rate (%)             :    59.01
Total Days               :  2381.00
Information Ratio        :     0.06
Tracking Error (%)       :    13.09

🔍 TEST A vs BASELINE COMPARISON:
Metric                    Baseline   Test A     Change    
-------------------------------------------

EXECUTIVE SUMMARY: Test A (Market Regime Overlay) shows mixed 
  results - significant drawdown improvement (+9.6%) but still
  exceeds institutional target. Sharpe ratio preserved (1.25 > 1.2)
  but Calmar ratio deteriorated due to return sacrifice. Need Test B
  and C to find optimal mechanism.

  DETAILED ANALYSIS:

  Test A Results Assessment:

  ✅ Positive Outcomes:
  - Drawdown Reduction: -39.6% vs -49.2% baseline (+9.6% improvement)
  - Sharpe Preservation: 1.25 > 1.2 institutional requirement ✅
  - Volatility Control: 10.30% vs 13.32% baseline (significant risk
  reduction)
  - Operational Success: Applied 50% exposure reduction to 43.8% of
  trading days

  ⚠️ Areas of Concern:
  - Target Miss: -39.6% still exceeds -25% institutional target by
  14.6%
  - Return Sacrifice: 12.83% vs 18.70% baseline (-5.87% annual return
   loss)
  - Calmar Deterioration: 0.32 vs 0.38 baseline (net risk-adjusted
  performance decline)

  📊 Key Insights:
  - Risk Coverage: 43.8% Bear/Stress periods provide substantial
  intervention opportunities
  - Mechanism Effectiveness: 50% exposure reduction during risk
  periods works but may be too conservative
  - Trade-off Balance: Substantial risk reduction comes at
  significant return cost

  IMPLEMENTATION NOTES:

  Test A validates the regime overlay approach but suggests
  calibration needs:
  1. Fine-tuning Opportunity: 50% reduction may be excessive - could
  test 60-70% exposure
  2. Institutional Gap: 14.6% gap to -25% target requires additional
  risk management
  3. Return Preservation: Need mechanism that maintains more alpha
  while reducing risk

In [8]:
# =================================================================
# CELL 6: VOLATILITY TARGETING OVERLAY IMPLEMENTATION (TEST B)
# =================================================================

def calculate_portfolio_volatility(
    portfolio_returns: pd.Series,
    window: int = 60
) -> pd.Series:
    """
    Calculate rolling portfolio volatility (annualized).
    """
    rolling_vol = portfolio_returns.rolling(window).std() * np.sqrt(252)
    return rolling_vol


def apply_volatility_targeting_overlay(
    baseline_holdings: pd.DataFrame,
    baseline_returns: pd.Series,
    target_volatility: float = 0.15,
    vol_window: int = 60,
    max_leverage: float = 1.0,
    min_exposure: float = 0.2
) -> pd.DataFrame:
    """
    Apply volatility targeting overlay by scaling exposure based on realized volatility.
    
    Parameters:
    - baseline_holdings: Daily portfolio weights from quarterly baseline
    - baseline_returns: Portfolio returns to calculate volatility
    - target_volatility: Target portfolio volatility (15%)
    - vol_window: Rolling window for volatility calculation (60 days)
    - max_leverage: Maximum exposure multiplier (1.0 = no leverage)
    - min_exposure: Minimum exposure multiplier (0.2 = 20% minimum)
    
    Returns:
    - vol_targeted_holdings: Adjusted portfolio weights with volatility targeting
    """
    print(f"🔧 Implementing Volatility Targeting Overlay (Test B)...")
    print(f"    Target volatility: {target_volatility:.1%}")
    print(f"    Volatility window: {vol_window} days")
    print(f"    Exposure range: {min_exposure:.1%} to {max_leverage:.1%}")

    # Calculate rolling portfolio volatility
    rolling_vol = calculate_portfolio_volatility(baseline_returns, window=vol_window)

    # Calculate volatility scaling factor
    vol_scaling = target_volatility / rolling_vol

    # Apply constraints
    vol_scaling = np.clip(vol_scaling, min_exposure, max_leverage)

    # Handle NaN values (early periods without enough data)
    vol_scaling = vol_scaling.fillna(1.0)

    # Apply scaling to holdings
    vol_targeted_holdings = baseline_holdings.copy()

    for date in vol_targeted_holdings.index:
        if date in vol_scaling.index:
            scaling_factor = vol_scaling[date]
            vol_targeted_holdings.loc[date] *= scaling_factor

    # Calculate statistics
    avg_scaling = vol_scaling[vol_scaling.notna()].mean()
    scaling_std = vol_scaling[vol_scaling.notna()].std()
    high_vol_days = (vol_scaling < 0.8).sum()  # Days with >80% reduction
    low_vol_days = (vol_scaling > 1.0).sum()    # Days with leverage (if any)

    print(f"    Average volatility scaling: {avg_scaling:.2f}")
    print(f"    Scaling volatility: {scaling_std:.2f}")
    print(f"    High volatility days (>80% reduction): {high_vol_days:,}")
    print(f"    Low volatility days (>100% exposure): {low_vol_days:,}")

    return vol_targeted_holdings, vol_scaling


# Execute Volatility Targeting Overlay (Test B)
print("\n" + "=" * 70)
print("🎯 IMPLEMENTING TEST B: VOLATILITY TARGETING OVERLAY")
print("=" * 70)

# Apply volatility targeting to quarterly baseline holdings
vol_target_holdings, vol_scaling_factors = apply_volatility_targeting_overlay(
    baseline_holdings=quarterly_holdings,
    baseline_returns=quarterly_returns,
    target_volatility=0.15,  # 15% target volatility
    vol_window=60,
    max_leverage=1.0,
    min_exposure=0.2
)

# Run backtest with volatility targeting
vol_target_returns, vol_target_log = run_regime_overlay_backtest(
    regime_holdings=vol_target_holdings,
    daily_returns=daily_returns_aligned,
    config=QUARTERLY_BASELINE_CONFIG,
    overlay_name="Volatility Targeting Overlay"
)

# Calculate performance metrics for Test B
vol_target_metrics = calculate_performance_metrics(vol_target_returns, benchmark_returns_aligned)

print("\n📊 VOLATILITY TARGETING OVERLAY PERFORMANCE (TEST B):")
print("=" * 50)
for metric, value in vol_target_metrics.items():
    if 'Ratio' in metric or 'Rate' in metric:
        print(f"{metric:<25}: {value:8.2f}")
    else:
        print(f"{metric:<25}: {value:8.2f}")

# Compare against baseline
print("\n🔍 TEST B vs BASELINE COMPARISON:")
print("=" * 40)
print(f"{'Metric':<25} {'Baseline':<10} {'Test B':<10} {'Change':<10}")
print("-" * 55)

for metric in key_metrics:
    baseline_val = quarterly_metrics[metric]
    overlay_val = vol_target_metrics[metric]
    change = overlay_val - baseline_val
    change_str = f"{change:+.2f}"
    print(f"{metric:<25} {baseline_val:8.2f}    {overlay_val:8.2f}    {change_str:>8s}")

# Risk assessment for Test B
print("\n🎯 TEST B RISK ASSESSMENT:")
print("=" * 30)

sharpe_preserved_b = vol_target_metrics['Sharpe Ratio'] >= 1.2
drawdown_improved_b = vol_target_metrics['Max Drawdown (%)'] > quarterly_metrics['Max Drawdown (%)']
calmar_improved_b = vol_target_metrics['Calmar Ratio'] > quarterly_metrics['Calmar Ratio']
target_achieved_b = vol_target_metrics['Max Drawdown (%)'] > -25.0

print(f"Sharpe Preservation (>1.2):      {'✅ PASS' if sharpe_preserved_b else '❌ FAIL'} ({vol_target_metrics['Sharpe Ratio']:.2f})")
print(f"Drawdown Improvement:            {'✅ YES' if drawdown_improved_b else '❌ NO'} ({vol_target_metrics['Max Drawdown (%)']:.1f}% vs {quarterly_metrics['Max Drawdown (%)']:.1f}%)")
print(f"Calmar Improvement:              {'✅ YES' if calmar_improved_b else '❌ NO'} ({vol_target_metrics['Calmar Ratio']:.2f} vs {quarterly_metrics['Calmar Ratio']:.2f})")
print(f"Target Achieved (<-25%):         {'✅ YES' if target_achieved_b else '❌ NO'} ({vol_target_metrics['Max Drawdown (%)']:.1f}%)")

if target_achieved_b and sharpe_preserved_b:
    print("\n🎉 TEST B SUCCESS: Volatility Targeting achieves institutional targets!")
elif drawdown_improved_b and sharpe_preserved_b:
    print("\n✅ TEST B PROGRESS: Significant improvement, may need fine-tuning")
else:
    print("\n⚠️ TEST B MIXED: Review trade-offs between volatility control and alpha preservation")

print("\n✅ Volatility Targeting Overlay (Test B) analysis complete")


🎯 IMPLEMENTING TEST B: VOLATILITY TARGETING OVERLAY
🔧 Implementing Volatility Targeting Overlay (Test B)...
    Target volatility: 15.0%
    Volatility window: 60 days
    Exposure range: 20.0% to 100.0%
    Average volatility scaling: 0.94
    Scaling volatility: 0.12
    High volatility days (>80% reduction): 341
    Low volatility days (>100% exposure): 0
🚀 Running Volatility Targeting Overlay backtest...
✅ Volatility Targeting Overlay backtest complete.

📊 VOLATILITY TARGETING OVERLAY PERFORMANCE (TEST B):
Annual Return (%)        :    17.62
Annual Volatility (%)    :    11.84
Sharpe Ratio             :     1.49
Max Drawdown (%)         :   -41.71
Calmar Ratio             :     0.42
Sortino Ratio            :     1.54
Win Rate (%)             :    59.18
Total Days               :  2381.00
Information Ratio        :     0.44
Tracking Error (%)       :    11.58

🔍 TEST B vs BASELINE COMPARISON:
Metric                    Baseline   Test B     Change    
------------------------------

EXECUTIVE SUMMARY: Test B (Volatility Targeting) delivers superior 
  performance - best Calmar ratio (0.42), improved Sharpe (1.49),
  minimal return sacrifice (-1.08%), and +7.5% drawdown improvement.
  Currently the leading risk management mechanism, though still 16.7%
   from institutional target.

  DETAILED ANALYSIS:

  Test B Results Assessment:

  ✅ Excellent Performance:
  - Calmar Ratio Leader: 0.42 vs 0.38 baseline (+0.04 improvement) 🏆
  - Sharpe Enhancement: 1.49 vs 1.40 baseline (+0.08 improvement)
  - Return Preservation: 17.62% vs 18.70% baseline (only -1.08%
  sacrifice)
  - Risk Reduction: -41.7% vs -49.2% baseline (+7.51% improvement)
  - Volatility Control: 11.84% vs 13.32% baseline (effective
  targeting)

  📊 Mechanism Effectiveness:
  - Dynamic Scaling: Average 0.94 scaling factor (6% average
  reduction)
  - Precision Targeting: 341 high-volatility days (14.3%) received
  >80% exposure reduction
  - Conservative Approach: Zero leverage days (respects 100% maximum
  exposure)

  ⚠️ Remaining Challenge:
  - Institutional Gap: -41.7% vs -25% target (16.7% gap remaining)

  🏆 Test B vs Test A Comparison:

  | Metric              | Test A         | Test B         | Winner |
  |---------------------|----------------|----------------|--------|
  | Return Preservation | 12.83% (-5.87) | 17.62% (-1.08) | Test B |
  | Sharpe Ratio        | 1.25           | 1.49           | Test B |
  | Max Drawdown        | -39.6%         | -41.7%         | Test A |
  | Calmar Ratio        | 0.32           | 0.42           | Test B |

  Key Insight: Volatility targeting provides more nuanced risk
  management than blunt regime-based exposure cuts.

  IMPLEMENTATION NOTES:

  Test B validates dynamic volatility targeting as superior to static
   regime overlays:
  1. Precision Approach: Responds to actual portfolio risk rather
  than market regimes
  2. Return Efficiency: Minimal alpha sacrifice for substantial risk
  reduction
  3. Institutional Quality: Best risk-adjusted performance (Calmar
  ratio)